src/interpreted/parser.py

```diff
@@ -302,6 +313,91 @@ def parse_single_line_statement(
         self.expect(TokenType.NEWLINE, TokenType.EOF)
         return node
 
+    def parse_import(self) -> Import:
+        self.expect(TokenType.NAME)
+
+        names = []
+        module = self.current().string
+
+        while True:
+            if self.match_op(","):
+                self.expect(TokenType.NAME)
+                names.append(alias(name=module, asname=None))
+                module = self.current().string
+                continue
+
+            elif self.match_name("as"):
+                self.expect(TokenType.NAME)
+                alias_name = self.current().string
+                names.append(alias(name=module, asname=alias_name))
+                if self.peek().token_type in (TokenType.NEWLINE, TokenType.EOF):
+                    break
+                elif self.match_op(","):
+                    self.expect(TokenType.NAME)
+                    module = self.current().string
+                continue
+
+            if self.match_op("."):
+                self.expect(TokenType.NAME)
+                module += "." + self.current().string
+                continue
+
+            if self.peek().token_type in (TokenType.NEWLINE, TokenType.EOF):
+                names.append(alias(name=module, asname=None))
+                break
+
+        return Import(names=names)
+
+    def parse_importfrom(self) -> ImportFrom:
+        self.expect(TokenType.NAME)
+
+        module_name = self.current().string
+        names = []
+
+        # parse submodule names or direct import keyword
+        if self.match_op("."):
+            while not self.match_name("import"):
+                if self.match_op("."):
+                    continue
+                self.expect(TokenType.NAME)
+                module_name += "." + self.current().string
+
+        elif not self.match_name("import"):
```

Comment:
```suggestion
        if not self.match_name("import"):
```

Reply:
The while loop that runs until `import` is reached is itself triggered by a `.`, so if I make this `if` here, the program breaks because it has moved past the `import` token at that point.

Reply:
You should rewrite the previous `while` loop as:

```diff
-        if self.match_op("."):
-            while not self.match_name("import"):
+        while not self.match_op("."):
```

Reply:
No need to match on `import` in the while loop. As long as there isn't a `.` it should break.

Reply:
Take inspiration from the grammar:
```
module -> NAME ('.' NAME)*
```